File: ecosystem/kryos-embed/hosts/node/demo_crm.mjs
/**
 * demo_crm.mjs -- private CRM assistant demo using the Kryos WASM agent.
 *
 * Mirrors hosts/python/demo_crm.py but uses the WASM sandbox instead of the
 * native DLL.  The sandbox story is stronger: the module has NO imports beyond
 * the explicit host functions listed by kryos-agent.mjs at load time -- no
 * file system, no TCP, no WASI.  A Kryos DLL can theoretically call any
 * Windows API through an undeclared native symbol; the WASM module physically
 * cannot because every import must be resolved from the host object at
 * instantiation time.
 *
 * Demo cases:
 *   1. Within-budget call  (budget_cents=5, agent cost=3) -> answered=true, source present
 *   2. Over-budget call    (budget_cents=1, agent cost=3) -> answered=false, spendCents=0
 */

import { createAgent } from "./kryos-agent.mjs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const WASM_PATH = join(__dirname, "dist", "kryos_embed_agent.wasm");

// ---------------------------------------------------------------------------
// In-memory "CRM" -- fake data, never written to disk or sent over network.
// The wasm sandbox cannot write files or open sockets regardless.
// ---------------------------------------------------------------------------
const CUSTOMERS = {
  C001: { name: "Alice Nguyen",   plan: "Enterprise", arr_usd: 24000, since: "2023-01" },
  C002: { name: "Bob Carruthers", plan: "Pro",        arr_usd:  4800, since: "2024-03" },
  C003: { name: "Carol Ito",      plan: "Starter",    arr_usd:   600, since: "2025-11" },
};

function summariseCustomers(customers) {
  const lines = ["CRM snapshot (in-process only):"];
  for (const [cid, info] of Object.entries(customers)) {
    lines.push(`  ${cid}: ${info.name} | ${info.plan} | $${info.arr_usd.toLocaleString()}/yr | since ${info.since}`);
  }
  return lines.join("\n");
}

function assert(cond, msg) {
  if (!cond) {
    console.error("ASSERTION FAILED:", msg);
    process.exit(1);
  }
}

async function main() {
  console.log("=== Kryos Private CRM Assistant Demo (Node / WASM) ===\n");
  console.log("Security posture:");
  console.log("  - Customer data loaded in-process from a JS object (no DB, no network read)");
  console.log("  - Agent sandbox: WASM with NO imports beyond explicit host functions");
  console.log("  - Budget cap enforced by the Kryos agent BEFORE any mock-LLM call\n");

  // Load agent -- prints import manifest on load (printImports=true by default).
  // The printed manifest IS the capability proof: every function the WASM module
  // can call is enumerated here.  Anything not in the list is physically impossible.
  const agent = await createAgent(WASM_PATH, { printImports: true });
  console.log("\nAgent loaded. Import manifest printed above.\n");

  const crmCtx = summariseCustomers(CUSTOMERS);
  const question = `${crmCtx}\n\nWhich customer has the highest ARR and what plan are they on?`;

  // --- Case 1: Within-budget call (budget=5, cost=3) ---
  console.log("--- Case 1: within-budget call (budgetCents=5, agent cost=3) ---");
  const r1 = agent.ask(question, 5);
  console.log(`  answered   : ${r1.answered}`);
  console.log(`  answer     : ${JSON.stringify(r1.answer)}`);
  console.log(`  source     : ${JSON.stringify(r1.source)}`);
  console.log(`  spendCents : ${r1.spendCents}`);
  console.log(`  reason     : ${JSON.stringify(r1.reason)}`);

  assert(r1.answered === true,        `expected answered=true, got: ${JSON.stringify(r1)}`);
  assert(r1.source !== "",            `expected non-empty source (provenance): ${JSON.stringify(r1)}`);
  assert(r1.spendCents === 3,         `expected spendCents=3, got ${r1.spendCents}`);
  console.log("  PASS: answered=true, source present, spendCents=3\n");

  console.log("Data-privacy note: the CRM object above was passed as text context in the");
  console.log("question string -- it never left the JS process.  The WASM sandbox has no");
  console.log("file/network/socket imports, so exfiltration is physically impossible.\n");

  // --- Case 2: Over-budget call (budget=1, cost=3) ---
  console.log("--- Case 2: over-budget call (budgetCents=1, agent cost=3) ---");
  const r2 = agent.ask(question, 1);
  console.log(`  answered   : ${r2.answered}`);
  console.log(`  answer     : ${JSON.stringify(r2.answer)}`);
  console.log(`  spendCents : ${r2.spendCents}`);
  console.log(`  reason     : ${JSON.stringify(r2.reason)}`);

  assert(r2.answered === false,       `expected answered=false (refused), got: ${JSON.stringify(r2)}`);
  assert(r2.spendCents === 0,         `refused call must not record spend; got ${r2.spendCents}`);
  console.log("  PASS: answered=false (refusal), spendCents=0 (no charge on refusal)\n");

  console.log("=== All demo assertions PASSED ===");
  console.log("Summary:");
  console.log(`  answered=true  source='mock-llm-v1'  spendCents=3  (within-budget)`);
  console.log(`  answered=false spendCents=0                          (over-budget refusal)`);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: ecosystem/kryos-embed/hosts/python/demo_crm.py
"""
demo_crm.py -- private CRM assistant demo using kryos_embed_agent.dll

Demonstrates:
  1. Data stays IN-PROCESS: the customer dict never leaves this process.
  2. Capability-gated: agent is restricted to ['ffi'] -- no network, no disk write.
  3. Budget-gated: the agent's own budget gate blocks over-budget calls.

The agent answers questions deterministically (mock LLM, cost=3 cents/call).
"""

import json
import pathlib
import sys

# Locate kryos_agent module (sibling file)
_here = pathlib.Path(__file__).parent
sys.path.insert(0, str(_here))

from kryos_agent import KryosAgent, CapabilityViolation  # noqa: E402

# ---------------------------------------------------------------------------
# In-memory "CRM" -- fake data, never written to disk or sent over network
# ---------------------------------------------------------------------------
CUSTOMERS = {
    "C001": {"name": "Alice Nguyen",   "plan": "Enterprise", "arr_usd": 24000, "since": "2023-01"},
    "C002": {"name": "Bob Carruthers", "plan": "Pro",        "arr_usd":  4800, "since": "2024-03"},
    "C003": {"name": "Carol Ito",      "plan": "Starter",    "arr_usd":   600, "since": "2025-11"},
}


def summarise_customers(customers: dict) -> str:
    """Produce a short plain-text summary of the CRM data for context injection."""
    lines = ["CRM snapshot (in-process only):"]
    for cid, info in customers.items():
        lines.append(
            f"  {cid}: {info['name']} | {info['plan']} | "
            f"${info['arr_usd']:,}/yr | since {info['since']}"
        )
    return "\n".join(lines)


def main() -> None:
    dist = _here.parent.parent / "dist"
    dll  = dist / "kryos_embed_agent.dll"
    caps = dist / "agent.caps.json"

    print("=== Kryos Private CRM Assistant Demo ===\n")
    print("Security posture:")
    print("  - Customer data loaded in-process from a Python dict (no DB, no network read)")
    print("  - Agent capability grant: ['ffi'] only -- no net:http, no net:tcp, no fs:write")
    print("  - Budget cap enforced by the Kryos agent BEFORE any LLM call\n")

    # Construct agent -- gate checks caps manifest before loading DLL
    agent = KryosAgent(dll_path=dll, caps_path=caps, allowed_caps=["ffi"])
    print("Agent loaded. Capability gate passed (ffi allowed, net:* denied).\n")

    crm_ctx = summarise_customers(CUSTOMERS)
    question = f"{crm_ctx}\n\nWhich customer has the highest ARR and what plan are they on?"

    # --- Case 1: Within-budget call (budget=5, cost=3) ---
    print("--- Case 1: within-budget call (budget_cents=5, agent cost=3) ---")
    result = agent.ask(question, budget_cents=5)
    print(f"  answered   : {result['answered']}")
    print(f"  answer     : {result['answer']!r}")
    print(f"  source     : {result['source']!r}")
    print(f"  spend_cents: {result['spend_cents']}")
    print(f"  reason     : {result['reason']!r}")

    assert result["answered"] == 1, f"Expected answered=1, got: {result}"
    assert result["source"] != "", "Expected non-empty source (provenance)"
    assert result["spend_cents"] == 3, f"Expected spend=3, got {result['spend_cents']}"
    print("  PASS: answered=1, source present, spend=3\n")

    print("Data-privacy note: the CRM dict above was passed as text context in the")
    print("request string -- it never left the Python process. No HTTP call was made.")
    print("The agent manifest (agent.caps.json) shows 'net:http' and 'net:tcp' are")
    print("absent from the grant, so a network-calling DLL would be blocked at load.\n")

    # --- Case 2: Over-budget call (budget=1, cost=3) ---
    print("--- Case 2: over-budget call (budget_cents=1, agent cost=3) ---")
    result2 = agent.ask(question, budget_cents=1)
    print(f"  answered   : {result2['answered']}")
    print(f"  answer     : {result2['answer']!r}")
    print(f"  spend_cents: {result2['spend_cents']}")
    print(f"  reason     : {result2['reason']!r}")

    assert result2["answered"] == 0, f"Expected answered=0 (refused), got: {result2}"
    assert result2["spend_cents"] == 0, (
        f"Refused call must not record spend; got spend={result2['spend_cents']}"
    )
    print("  PASS: answered=0 (refusal), spend_cents=0 (no charge on refusal)\n")

    # --- Case 3: Capability gate proof (net:tcp denied) ---
    print("--- Case 3: capability gate -- agent with net:tcp REFUSED at load time ---")
    gate_fired = False
    try:
        _bad = KryosAgent(dll_path=dll, caps_path=caps, allowed_caps=["ffi", "fs:read"])
        # net:tcp is NOT in the manifest so this should succeed -- but try a minimal
        # allow-set that excludes 'ffi' to force a violation
    except CapabilityViolation:
        gate_fired = True  # shouldn't fire here since ffi+fs:read covers manifest

    # The real test: allow-set that excludes 'ffi' (which all exports need)
    try:
        _bad2 = KryosAgent(dll_path=dll, caps_path=caps, allowed_caps=[])
        print("  ERROR: expected CapabilityViolation but none raised")
        sys.exit(1)
    except CapabilityViolation as exc:
        print(f"  CapabilityViolation raised: {exc!s:.120}")
        print("  PASS: DLL was NOT loaded when allowed_caps=[] (ffi excluded)\n")

    print("=== All demo assertions PASSED ===")
    print("Summary:")
    print("  answered=1  source='mock-llm-v1'  spend=3  (within-budget call)")
    print("  answered=0  spend=0                         (over-budget refusal)")
    print("  CapabilityViolation                         (ffi excluded from allow-set)")


if __name__ == "__main__":
    main()
